migrations/2025-04-25-090000_trusted-publishing/up.sql

```diff
@@ -0,0 +1,64 @@
+create table github_oidc_configs
+(
+    id serial primary key,
+    created_at timestamptz not null default now(),
+    crate_id int not null references crates on delete cascade,
+    repository_owner varchar not null,
+    repository_owner_id int not null,
+    repository_name varchar not null,
+    workflow_filename varchar not null,
+    environment varchar
+);
+
+comment on table github_oidc_configs is 'Trusted Publisher OIDC configuration for GitHub Actions';
+comment on column github_oidc_configs.id is 'Unique identifier of the `github_oidc_configs` row';
+comment on column github_oidc_configs.created_at is 'Date and time when the configuration was created';
+comment on column github_oidc_configs.crate_id is 'Unique identifier of the crate that this configuration is for';
+comment on column github_oidc_configs.repository_owner is 'GitHub name of the user or organization that owns the repository';
+comment on column github_oidc_configs.repository_owner_id is 'GitHub ID of the user or organization that owns the repository';
+comment on column github_oidc_configs.repository_name is 'Name of the repository that this configuration is for';
+comment on column github_oidc_configs.workflow_filename is 'Name of the workflow file inside the repository that will be used to publish the crate';
+comment on column github_oidc_configs.environment is 'GitHub Actions environment that will be used to publish the crate (if `NULL` the environment is unrestricted)';
+
+create index github_oidc_configs_repository_owner_id_repository_name_index
+    on github_oidc_configs (repository_owner_id, repository_name);
+
+-------------------------------------------------------------------------------
+
+create table oidc_tokens
+(
+    id bigserial primary key,
+    created_at timestamptz not null default now(),
+    crate_id int not null references crates on delete cascade,
+    hashed_token bytea not null,
+    expires_at timestamptz not null
+);
+
+comment on table oidc_tokens is 'Temporary access tokens for OIDC-based publishing (aka. Trusted Publishing)';
+comment on column oidc_tokens.id is 'Unique identifier of the `oidc_tokens` row';
+comment on column oidc_tokens.created_at is 'Date and time when the token was created';
+comment on column oidc_tokens.crate_id is 'Unique identifier of the crate that can be published using this token';
+comment on column oidc_tokens.hashed_token is 'SHA256 hash of the token that can be used to publish the crate';
```

Comment:
a JWT can generally be used to publish multiple crates, if they all have a similar Trusted Publishing configuration for the same repository. during the exchange request the API server would create a dedicated `oidc_tokens` row for each crate with the same `hashed_token`.

would it be better to use a `crate_ids` array instead of row-per-crate? I'm not sure if an array would be able to support the `references crates` above.

on the other hand, the current design does not prevent hash collisions since a uniqueness constraint on only the `hashed_token` column is not possible.

Reply:
Yeah, I forget how arrays work when used as foreign keys. Seems simpler to leave it the way you have it here with multiple rows and no constraint on the `hashed_token`.

Reply:
> Yeah, I forget how arrays work when used as foreign keys

apparently they are not implemented at all according to the internet, but since the tokens are short-lived anyway it's probably not a huge deal to have IDs in the array that no longer correspond to an existing crate. since the crate IDs are unique and won't get reassigned we shouldn't be vulnerable to resurrection attacks or anything like that.

I've now adjusted the PR and the prototype to use the one-row-per-token approach with a `crate_ids: int[]` column, which then gives us the extra safety regarding hash collisions.